<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from /Users/yury/workspaces/git/youbot-ros-pkg/youbot_common/youbot_description/robots/youbot_arm.urdf.xacro | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="youbot" xmlns:controller="http://playerstage.sourceforge.net/gazebo/xmlschema/#controller" xmlns:interface="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface" xmlns:sensor="http://playerstage.sourceforge.net/gazebo/xmlschema/#sensor" xmlns:xacro="http://ros.org/wiki/xacro">
  <!-- The following included files set up definitions of parts of the robot body -->
  <!-- misc common stuff? -->
  <!-- materials for visualization -->
  <material name="Orange">
    <color rgba="1.0 0.4 0.0 1.0"/>
  </material>
  <material name="Grey">
    <color rgba="0.7 0.7 0.7 1.0"/>
  </material>
  <material name="DarkGrey">
    <color rgba="0.3 0.3 0.3 1.0"/>
  </material>
  <material name="White">
    <color rgba="1.0 1.0 1.0 1.0"/>
  </material>
  <material name="Black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="Red">
    <color rgba="1.0 0.0 0.0 1.0"/>
  </material>
  <material name="Green">
    <color rgba="0.0 1.0 0.0 1.0"/>
  </material>
  <!-- youbot arm -->
  <!-- youbot arm -->
  <!-- generic simulator_gazebo plugins for starting mechanism control, ros time, ros battery -->
  <gazebo>
    <controller:gazebo_ros_controller_manager name="gazebo_ros_controller_manager" plugin="libgazebo_ros_controller_manager.so">
      <alwaysOn>true</alwaysOn>
      <updateRate>1000.0</updateRate>
      <interface:audio name="gazebo_ros_controller_manager_dummy_iface"/>
    </controller:gazebo_ros_controller_manager>
    <controller:gazebo_ros_power_monitor name="gazebo_ros_power_monitor_controller" plugin="libgazebo_ros_power_monitor.so">
      <alwaysOn>true</alwaysOn>
      <updateRate>1.0</updateRate>
      <timeout>5</timeout>
      <interface:audio name="power_monitor_dummy_interface"/>
      <powerStateTopic>power_state</powerStateTopic>
      <powerStateRate>10.0</powerStateRate>
      <fullChargeCapacity>87.78</fullChargeCapacity>
      <dischargeRate>-474</dischargeRate>
      <chargeRate>525</chargeRate>
      <dischargeVoltage>15.52</dischargeVoltage>
      <chargeVoltage>16.41</chargeVoltage>
    </controller:gazebo_ros_power_monitor>
  </gazebo>
  <!-- Now we can start using the macros included above to define the actual youbot -->
  <!-- The first use of a macro.  This one was defined in base.urdf.xacro above.
       A macro like this will expand to a set of link and joint definitions, and to additional
       Gazebo-related extensions (sensor plugins, etc).  The macro takes an argument, name, 
       that equals "base", and uses it to generate names for its component links and joints 
       (e.g., base_link).  The included origin block is also an argument to the macro.  By convention, 
       the origin block defines where the component is w.r.t its parent (in this case the parent 
       is the world frame). For more, see http://www.ros.org/wiki/xacro -->
  <!-- foot for arm-->
  <link name="base_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="10.0"/>
      <inertia ixx="0.1" ixy="0" ixz="0" iyy="0.1" iyz="0" izz="0.1"/>
    </inertial>
  </link>
  <joint name="arm_joint_0" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="base_link"/>
    <child link="arm_link_0"/>
  </joint>
  <link name="arm_link_0">
    <inertial>
      <!--<origin xyz="0.023 0 0.006" rpy="0 0 0"/>-->
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.845"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_arm/arm0.dae"/>
      </geometry>
      <material name="DarkGrey"/>
    </visual>
    <collision>
      <!--<origin xyz="0.023 0 0.006" rpy="0 0 0"/>-->
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_arm/arm0_convex.dae"/>
        <!--<cylinder length="0.078" radius="0.093"/>-->
      </geometry>
    </collision>
  </link>
  <joint name="arm_joint_1" type="revolute">
    <origin rpy="0 0 2.96705972839" xyz="0.024 0 0.096"/>
    <parent link="arm_link_0"/>
    <child link="arm_link_1"/>
    <axis xyz="0 0 -1"/>
    <!--<calibration rising="${arm_1_ref}" falling="${arm_1_ref}"/>-->
    <dynamics damping="0.5" friction="0.5"/>
    <!-- damping - N*s/m, N*s*m/rad -->
    <!-- friction - N, N*m -->
    <limit effort="1" lower="0" upper="5.89921287174" velocity="1.57079632679"/>
    <!--effort - N, N*m-->
    <!--velocity - m/s, rad/s-->
    <safety_controller k_position="20" k_velocity="100" soft_lower_limit="0.001" soft_upper_limit="5.89821287174"/>
    <!--k_position N/m, N/rad-->
    <!--k_velocity N*s/m, N*s*m/rad-->
    <!--for details: http://www.ros.org/wiki/pr2_controller_manager/safety_limits-->
  </joint>
  <link name="arm_link_1">
    <inertial>
      <mass value="2.412"/>
      <origin rpy="3.14159265359  0.349065850399 0.0" xyz="0.01489 0.00213 0.00213"/>
      <inertia ixx="0.0035070" ixy="0" ixz="0" iyy="0.0065526" iyz="0" izz="0.0063686"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_arm/arm1.dae"/>
      </geometry>
      <!--<material name="Orange" />-->
    </visual>
    <collision>
      <!--<origin xyz="0 0 -0.003" rpy="0 0 0" />-->
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_arm/arm1_convex.dae"/>
        <!--<cylinder length="0.102" radius="0.080"/>-->
      </geometry>
    </collision>
  </link>
  <joint name="arm_joint_2" type="revolute">
    <origin rpy="0 -1.1344640138 0" xyz="0.033 0 0.019"/>
    <parent link="arm_link_1"/>
    <child link="arm_link_2"/>
    <axis xyz="0 1 0"/>
    <!--<calibration rising="${arm_2_ref}" falling="${arm_2_ref}"/>-->
    <dynamics damping="0.5" friction="0.5"/>
    <limit effort="10" lower="0" upper="2.70526034059" velocity="1.57079632679"/>
    <safety_controller k_position="20" k_velocity="1.0" soft_lower_limit="0.001" soft_upper_limit="2.70426034059"/>
  </joint>
  <link name="arm_link_2">
    <inertial>
      <mass value="1.155"/>
      <origin rpy="-1.57079632679  0.0 -1.57079632679" xyz="0.10453 0.0017 -0.00297"/>
      <inertia ixx="0.0044395" ixy="0" ixz="0" iyy="0.0008231" iyz="0" izz="0.004447"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 -0.032 0.078"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_arm/arm2.dae"/>
      </geometry>
      <material name="Orange"/>
    </visual>
    <collision>
      <!--<origin xyz="0.001 -0.041 0.074" rpy="0 0 0" />-->
      <origin rpy="0 0 0" xyz="0 -0.032 0.078"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_arm/arm2_convex.dae"/>
        <!--<box size="0.074 0.082 0.222"/>-->
      </geometry>
    </collision>
  </link>
  <joint name="arm_joint_3" type="revolute">
    <origin rpy="0 2.54818070791 0" xyz="0.000 0.000 0.155"/>
    <parent link="arm_link_2"/>
    <child link="arm_link_3"/>
    <axis xyz="0 1 0"/>
    <!--<calibration rising="${arm_3_ref}" falling="${arm_3_ref}"/>-->
    <dynamics damping="0.5" friction="0.5"/>
    <limit effort="1" lower="-5.18362787842" upper="0" velocity="1.57079632679"/>
    <safety_controller k_position="20" k_velocity="1.0" soft_lower_limit="-5.18262787842" soft_upper_limit="0.001"/>
  </joint>
  <link name="arm_link_3">
    <inertial>
      <mass value="0.934"/>
      <origin rpy="0.0  0.0 1.57079632679" xyz="0.0001 0.0967 0.02107"/>
      <inertia ixx="0.0024588" ixy="0" ixz="0" iyy="0.0005354" iyz="0" izz="0.0025708"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0.000 0.028 0.079"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_arm/arm3.dae"/>
      </geometry>
      <material name="Orange"/>
    </visual>
    <collision>
      <!--<origin xyz="0.000 0.025 0.067" rpy="0 0 0" />-->
      <origin rpy="0 0 0" xyz="0.000 0.028 0.079"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_arm/arm3_convex.dae"/>
        <!--<box size="0.064 0.054 0.192"/>-->
      </geometry>
    </collision>
  </link>
  <joint name="arm_joint_4" type="revolute">
    <origin rpy="0 -1.78896248329 0" xyz="0.000 0.000 0.135"/>
    <parent link="arm_link_3"/>
    <child link="arm_link_4"/>
    <axis xyz="0 1 0"/>
    <!--<calibration rising="${arm_4_ref}" falling="${arm_4_ref}"/>-->
    <dynamics damping="0.5" friction="0.5"/>
    <limit effort="1" lower="0" upper="3.57792496659" velocity="1.57079632679"/>
    <safety_controller k_position="20" k_velocity="1.0" soft_lower_limit="0" soft_upper_limit="3.57692496659"/>
  </joint>
  <link name="arm_link_4">
    <inertial>
      <mass value="0.877"/>
      <origin rpy="0.0  3.14159265359 0.698131700798" xyz="0.00011 0.05103 -0.02319"/>
      <inertia ixx="0.0008688" ixy="0" ixz="0" iyy="0.0013669" iyz="0" izz="0.0008970"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 -0.010 0.029"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_arm/arm4.dae"/>
      </geometry>
      <material name="Orange"/>
    </visual>
    <collision>
      <!--<origin xyz="0 0 0.026" rpy="0 0 0" />-->
      <origin rpy="0 0 0" xyz="0 -0.010 0.029"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_arm/arm4_convex.dae"/>
        <!--<box size="0.058 0.100 0.118"/>-->
      </geometry>
    </collision>
  </link>
  <joint name="arm_joint_5" type="revolute">
    <origin rpy="0 0 2.92342649709" xyz="-0.002 0 0.130"/>
    <parent link="arm_link_4"/>
    <child link="arm_link_5"/>
    <axis xyz="0 0 1"/>
    <!--<calibration rising="${arm_5_ref}" falling="${arm_5_ref}"/>-->
    <dynamics damping="0.5" friction="0.5"/>
    <limit effort="1" lower="0" upper="5.84685299418" velocity="1.57079632679"/>
    <safety_controller k_position="20" k_velocity="1.0" soft_lower_limit="0" soft_upper_limit="5.84585299418"/>
  </joint>
  <link name="arm_link_5">
    <inertial>
      <mass value="0.251"/>
      <origin rpy="0.0  1.57079632679 0.0" xyz="0.0 0.00115 -0.01683"/>
      <inertia ixx="0.0003388" ixy="0" ixz="0" iyy="0.0002798" iyz="0" izz="0.0001191"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0.003 0 -0.034"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_arm/arm5.dae"/>
      </geometry>
      <material name="DarkGrey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0.003 0 -0.034"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_arm/arm5_convex.dae"/>
        <!--<box size="0.054 0.096 0.030"/>-->
      </geometry>
    </collision>
  </link>
  <gazebo reference="arm_link_0">
    <material value="youbot/DarkGrey"/>
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <gazebo reference="arm_link_1">
    <!--<material value="youbot/Orange" />-->
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <gazebo reference="arm_link_2">
    <!--<material value="youbot/Orange" />-->
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <gazebo reference="arm_link_3">
    <!--<material value="youbot/Orange" />-->
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <gazebo reference="arm_link_4">
    <!--<material value="youbot/Orange" />-->
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <gazebo reference="arm_link_5">
    <!--<material value="youbot/DarkGrey" />-->
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <transmission name="arm_trans_1" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="arm_motor_1"/>
    <joint name="arm_joint_1"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>
  <transmission name="arm_trans_2" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="arm_motor_2"/>
    <joint name="arm_joint_2"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>
  <transmission name="arm_trans_3" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="arm_motor_3"/>
    <joint name="arm_joint_3"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>
  <transmission name="arm_trans_4" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="arm_motor_4"/>
    <joint name="arm_joint_4"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>
  <transmission name="arm_trans_5" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="arm_motor_5"/>
    <joint name="arm_joint_5"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>
  <joint name="gripper_palm_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="arm_link_5"/>
    <child link="gripper_palm_link"/>
  </joint>
  <link name="gripper_palm_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_gripper/palm.dae"/>
      </geometry>
      <material name="DarkGrey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <!--<box size="0.050 0.094 0.054"/>-->
        <mesh filename="package://youbot_description/meshes/youbot_gripper/palm_convex.dae"/>
      </geometry>
    </collision>
  </link>
  <joint name="gripper_finger_joint_l" type="prismatic">
    <origin rpy="0 0 3.14159265359" xyz="0.004 0.008 0.046"/>
    <parent link="gripper_palm_link"/>
    <child link="gripper_finger_link_l"/>
    <axis xyz="0 -1 0"/>
    <dynamics damping="1.0" friction="1.0"/>
    <limit effort="1" lower="0" upper="0.0115" velocity="0.1"/>
    <safety_controller k_position="20" k_velocity="25" soft_lower_limit="0.001" soft_upper_limit="0.011"/>
  </joint>
  <link name="gripper_finger_link_l">
    <inertial>
      <origin rpy="0 0 0" xyz="0 -0.001 0"/>
      <mass value="0.01"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_gripper/finger.dae"/>
      </geometry>
      <material name="DarkGrey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <!--<box size="0.014 0.020 0.046"/>-->
        <mesh filename="package://youbot_description/meshes/youbot_gripper/finger_convex.dae"/>
      </geometry>
    </collision>
  </link>
  <joint name="gripper_finger_joint_r" type="prismatic">
    <origin rpy="0 0 0" xyz="0.004 -0.008 0.046"/>
    <parent link="gripper_palm_link"/>
    <child link="gripper_finger_link_r"/>
    <axis xyz="0 -1 0"/>
    <dynamics damping="1.0" friction="1.0"/>
    <limit effort="1" lower="0" upper="0.0115" velocity="0.1"/>
    <safety_controller k_position="20" k_velocity="25" soft_lower_limit="0.001" soft_upper_limit="0.011"/>
  </joint>
  <link name="gripper_finger_link_r">
    <inertial>
      <origin rpy="0 0 0" xyz="0.000 -0.001 0"/>
      <mass value="0.01"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://youbot_description/meshes/youbot_gripper/finger.dae"/>
      </geometry>
      <material name="DarkGrey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0.000 0 0"/>
      <geometry>
        <!--<box size="0.014 0.020 0.046"/>-->
        <mesh filename="package://youbot_description/meshes/youbot_gripper/finger_convex.dae"/>
      </geometry>
    </collision>
  </link>
  <gazebo reference="gripper_palm_link">
    <material>Gazebo/Black</material>
    <turnGravityOff>false</turnGravityOff>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="gripper_finger_l_link">
    <material>Gazebo/Black</material>
    <turnGravityOff>false</turnGravityOff>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="gripper_finger_r_link">
    <material>Gazebo/Black</material>
    <turnGravityOff>false</turnGravityOff>
    <selfCollide>true</selfCollide>
  </gazebo>
  <transmission name="gripper_finger_l_trans" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="gripper_finger_l_motor"/>
    <joint name="gripper_finger_joint_l"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>
  <transmission name="gripper_finger_r_trans" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="gripper_finger_r_motor"/>
    <joint name="gripper_finger_joint_r"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>
</robot>
